### ISSUE TITLE
训练结果不稳定

### ISSUE THREAD

**Issue by RechardWong** [yzyouzhang/AIR-ASVspoof#11]
你好，我下载代码进行训练时发现相同参数重复跑两次所得的结果相差比较大。
对于ocsoftmax损失，最好的情况在eval上EER为2%，最差的情况是3%+，(其他损失函数也存在这种情况)
我也考虑了随机数的影响，但是训练到最后稳定下来正常情况应该是不会相差很大的。结果相差这么大是不是不正常，不知道你是否也有这样的情况。

**Comment by yzyouzhang** [comment]
感谢您对我们的工作感兴趣！
其他一些研究人员也注意到了这个问题，我们正在努力提高可重复性。最近我做了更多实验发现OC-Softmax和二分类损失之间的差异在统计上是显着的，从而验证了我们方法的有效性。
至于训练结果不稳定的原因，我可以想到以下两点：
1）dev集和eval集之间的分布不同。我们训练了100轮并根据dev集的性能选择模型。我们在eval集上观察到很多差异，但是在dev集上差异不明显。从这个意义上说，在dev集上基于EER选择模型并不是保证eval集性能良好的最佳方法。
2）模型的初始权重有很大的差异，因此在训练期间，它们收敛到不同的局部最小值。这些局部最小值在dev集上具有相似的性能，但在eval集上具有不同的性能。初始化可能会对ResNet产生很大影响。
如果您还有其他想法请联系我。谢谢！

Thank you for your interest in our work! 
Some other researchers also observe this issue, and we are trying to improve the reproducibility. I did more experiments recently; the difference between the OC-Softmax and binary classification losses is statistically significant, verifying our method's effectiveness.
As for the reasons why the results have much variance, I can think of:
1) the distribution is different between the development set and the evaluation set. We trained 100 epochs and chose the model based on the performance on the development set. We observed much variance on the evaluation set but not much on the development set. In this sense, selecting models based on EER on the development set is not the best way to guarantee good performance on the evaluation set.
2) The model's initial weights have much variance, so during training, they converge to different local minimums. These local minimums have similar performance on the development set but have different performance on the evaluation set. The initialization may have much influence on ResNet.
Please feel free to contact me if you have further thoughts. Thanks!

**Comment by RechardWong** [comment]
谢谢你的回复！
是的，你们的方法从原理和实验结果来看都非常棒，正是由于你们的工作非常出色，相对baseline好很多，即使实验结果不稳定也能证明你们的方法有效。假如方法的提升不明显的话，因为实验结果不稳定的原因就很难说明方法是否真的有效。

关于不稳定的原因，我的想法也跟你差不多，不过最终我更愿意归结于数据集的问题：
1、训练集和测试集分布不一致，train 和dev分布相近，所以在dev集上比较稳定，在eval上的性能差异会比较大。
2、训练集与测试集比例或许也有点影响，其他深度学习任务的数据集训练集是测试集的4~5倍，所以得到的结果会比较稳定，而这里相反，所以结果差异比较大（但是对于基于GMM的方法，结果相对稳定点。）。

另外对于样本特征处理，代码中训练阶段和测试阶段处理方式是相同的，长于7.5s随机选取一段连续的7.5s特征输入网络。这样在对于测试阶段，没有固定随机数的情况下同一个样本输入网络的特征可能不同，这或许会使结果出现一些随机性。然后我测试阶段固定选取前7.5s作为输入特征，但还是会出现上述不稳定的情况，说明这个不是主要原因。

```python
def setup_seed(seed):
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    np.random.seed(seed)
    random.seed(seed)
    torch.backends.cudnn.deterministic = True
```

目前也只能通过固定随机数种子的方式重现实验结果。我注意到代码中也有固定部分随机数，但是无法完全重现训练结果，通过上面代码固定随机数种子，相同参数训练结果是可重现的。

**Comment by yzyouzhang** [comment]
非常感谢你的回复！ 非常有帮助！
我同意您的看法，结果的波动可能是由于数据集性质造成的。 我们需要更多的实验来探讨这一点。
关于你说的长于7.5s随机选取连续一段的随机性，我们发现如果固定训好的模型，跑多次test代码的话，EER波动在0.01%以内，所以这一点对结果不稳定的影响确实很小。
您的设置随机数种子代码非常有帮助！ 可以重现相同超参数的训练结果。 
谢谢！

Thank you so much for your response! That's very helpful!
I agree with you that the fluctuation of the results may be due to the dataset. We need more experiments to investigate that.
Regarding the variance of randomly selecting a consecutive segment for trials longer than 7.5s, we found that if the trained model is fixed and we ran the test code multiple times, the EER fluctuates within 0.01%, so the randomness of randomly pick a consecutive piece does not contribute much to the overall variance.
Your code for setting up the random seed is pretty helpful! It works to reproduce the result with the same hyperparameters. Thanks!
